ComplianceAsCode/compliance-operator
# Add rule type change detection in profile parse

We went through migration cases for the new KubeletConfig changes. We decided to handle the TailoredProfile migration on the operator instead.

This fixes future rule checkType change migration issues as well as rule/variables deprecation issues.

To spot deprecated variables and rules, we'll use annotations derived from the content:

* Variable/Rule deprecation: `compliance.openshift.io/deprecated`

During the profileParser upgrade, we'll detect rule TypeChanges by comparing existing rule types against the new updates. If there's a change, the annotation `compliance.openshift.io/rule-last-check-type` will be added to the rule.

We will issues event and show warning in the TailoredProfile Status, if there are any deprecated or migrated rules/variables in the TailoredProfile, and user can also choose to set following annotation on the tailoredProfile: `compliance.openshift.io/prune-outdated-references`, so that the tailoredProfile controller will removal of all deprecated rules and variables along with migrated rules references from that specific tailoredProfile object.

To bypass migration and deprecation validation on specific TailoredProfile object, add the tag: `compliance.openshift.io/disable-outdated-reference-validation`.


Testing Profiles:

* Initial Setup:
   * Install CO prior to the release of the KubeletConfig patch.
   * Wait for all the profile bundles to reach a 'ready' state.
* Deploying TailoredProfile (TP):
   * Deploy the TP. This should lead to an 'all ready' state for TP.
* Modifying ProfileBundle Deployment:
  * Remove the profile bundle deployment. Important: Ensure that the profile bundle object remains intact and isn't deleted.
* Re-deploying CO with the Patch:
  * Make sure the operator pod has restarted and ProfileBundle deployment is being created
* Validation:
  * Verify that the TP displays the appropriate states.

Check [comment](https://github.com/ComplianceAsCode/compliance-operator/pull/401#issuecomment-1715418241) for testing profiles
